basic/src/shared/components/Dialog.tsx

```diff
@@ -0,0 +1,99 @@
+import React from "react";
+import { createPortal } from "react-dom";
+import { twJoin } from "tailwind-merge";
+import { usePortalContainer } from "../hooks/usePortalContainer";
+
+interface DialogProps extends React.PropsWithChildren {
+  open: boolean;
+  onClose: () => void;
+  closeOnClickOutside?: boolean;
+}
+
+export function PortalDialog(props: DialogProps) {
+  const container = usePortalContainer();
+
+  if (!container) {
+    return null;
+  }
+
+  return createPortal(<Dialog {...props} />, container);
+}
+
+function Dialog({
+  open,
+  onClose,
+  children,
+  closeOnClickOutside = true,
+}: DialogProps) {
+  const dialogRef = React.useRef<HTMLDialogElement>(null);
+
+  React.useEffect(
+    function handleShowOrCloseDialog() {
+      const dialog = dialogRef.current;
+      if (!dialog) return;
+
+      if (open && !dialog.open) {
+        dialog.showModal();
+      } else if (!open && dialog.open) {
+        dialog.close();
+      }
+    },
+    [open],
+  );
+
+  React.useEffect(
+    function handleCloseOnClickOutside() {
+      if (!closeOnClickOutside) return;
+
+      const dialog = dialogRef.current;
+      if (!dialog) return;
+
+      console.log(dialog);
+
+      const handleClick = (e: MouseEvent) => {
+        const rect = dialog.getBoundingClientRect();
+        const clickedOutside =
+          e.clientX < rect.left ||
+          e.clientX > rect.right ||
+          e.clientY < rect.top ||
+          e.clientY > rect.bottom;
+
+        if (clickedOutside) {
+          onClose();
+        }
+      };
+
+      dialog.addEventListener("click", handleClick);
+      return () => {
+        dialog.removeEventListener("click", handleClick);
+      };
+    },
+    [closeOnClickOutside, onClose],
+  );
+
+  React.useEffect(
+    function handlePreventScroll() {
+      if (!open) return;
+
+      const originalOverflow = document.body.style.overflow;
+      document.body.style.overflow = "hidden";
+      return () => {
+        document.body.style.overflow = originalOverflow;
+      };
+    },
+    [open],
+  );
+
+  return (
+    <dialog
+      ref={dialogRef}
+      className={twJoin(
+        "top-[20vh] my-0 flex max-h-[55vh] overflow-hidden",
```

Comment:
```suggestion
        "max-h-2xl flex overflow-hidden",
```

I'm not sure if you need the `top` and the `my` styles. Also, you can go for one of the common `max-w-*` values instead of the custom `55vh` height.

I'd remove `overflow-hidden` and the `shadow-lg` (since the child's shadow wouldn't be hidden anymore) and add `overflow-auto` to the child element.

Reply:
Unfortunately something like `max-h-2xl`, height doesn't get the same breakpoints as width.

My idea is for the dialog be centered on the X axis and a little above the center on the Y axis.
The dialog should always expand down, not both up and down when the height changes.

To achieve that we need to fix the top of the dialog. In that case `my-0` is necessary because `<dialog>` forces `m-auto`. That means that dialog will start at 20vh and then be centered in the left-over space.

So we have:
20vh blank
[0-55]vh content
[25-75]vh blank

I agree on the overflow and shadow part.